### PR TITLE
Shift years

### PR DESCRIPTION
**Pull request by zlister**
This PR is adding the `shift_years` function so that the dates on the patients will be adjusted according to the new measurement period.

- [x] Update cqm-models in the Gemfile once merged

Pull requests into Bonnie require the following. Submitter and reviewer should :white_check_mark: when done. For items that are not-applicable, note it's not-applicable ("N/A") and :white_check_mark:.

**Submitter:**
- [x] This pull request describes why these changes were made.
- [x] This PR is into the correct branch.
- [x] JIRA ticket for this PR: https://jira.mitre.org/browse/BONNIE-1848
- [x] JIRA ticket links to this PR
- [x] Code diff has been done and been reviewed (it **does not** contain: additional white space, not applicable code changes, debug statements, etc.)
- [x] Tests are included and test edge cases
- [x] Tests have been run locally and pass (remember to update Gemfile when applicable)
- [x] Test fixtures updated and documented as necessary ( see [internal wiki](https://gitlab.mitre.org/bonnie/internal-documentation/wikis/testing#test-fixtures) )
- [x] Code coverage has not gone down and all code touched or added is covered. 
     * In rare situations, this may not be possible or applicable to a PR. In those situations:
         1. Note why this could not be done or is not applicable here: 
         2. Add TODOs in the code noting that it requires a test
         3. Add a JIRA task to add the test and link it here: 
- [x] Automated regression test(s) pass

**Reviewer 1:**

Name:
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA tests have been run and pass
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree

**Reviewer 2:**

Name: @losborne 
- [x] Code is maintainable and reusable, reuses existing code and infrastructure where appropriate, and accomplishes the task’s purpose
- [x] The tests appropriately test the new code, including edge cases
- [x] You have tried to break the code

If JIRA tests were used to supplement or replace automated tests:
- [x] JIRA tests have been run and pass
- [x] You agree with the justification for use of JIRA tests or have provided input on why you disagree

### CODE REVIEW

File: app/controllers/measures_controller.rb [hunk 1]
```diff
@@ -117,4 +120,12 @@ def retrieve_measure_details(params)
     }
   end
 
+  def shift_years(measure, year)
+    CQM::Patient.by_user_and_hqmf_set_id(current_user, measure.hqmf_set_id).each do |patient|
+      patient.qdmPatient.dataElements.each do |data_element|
```
Comment (lukoktonos) [inline app/controllers/measures_controller.rb]: In addition to this loop, we should also update the birthDatetime of the qdmpatient (which is in the data elements array but also outside of it:

image

File: test/functional/measures_controller_test.rb [hunk 2]
```diff
@@ -760,15 +760,20 @@ class << update_measure_file
...
       year: '1984',
-      id: measure.id.to_s
+      id: measure.id.to_s,
+      measurement_period_shift_dates: "true"
     }
     measure = CQM::Measure.where(id: measure_id).first
     assert_equal '1984', measure.measure_period['low']['value'].slice(0,4)
+    patient = CQM::Patient.by_user(@user).first
+    assert_equal 1984, patient.qdmPatient.dataElements.first.authorDatetime.year
+    assert_equal 1984, patient.qdmPatient.dataElements.first.relevantPeriod.low.year
```
Comment (lukoktonos) [inline test/functional/measures_controller_test.rb]: Should add a test where the data element had a year other than 2012 originally and check that it has shifted appropriately by the correct number of years

One edge case that is unlikely but worth considering is if the shift in dates would produce an invalid CQL Year,
for example if a data element has year 0001 but then the MP is changed from 2012 to 2011.  We might want to throw an error in that case

**Review by lukoktonos — CHANGES_REQUESTED** [review]

File: app/controllers/measures_controller.rb (hunk 1 above)
Comment (jbradl11) [reply app/controllers/measures_controller.rb]: In addition to this, would a death Datetime be handled in a similar way?

File: lib/tasks/fixture_helper.rb [hunk 3]
```diff
@@ -45,12 +45,27 @@ def load_fixture_file(file, collection_name, user = nil)
...
   return nil unless json.is_a?(Hash)
   json.each_pair do |k,v|
-    json[k] = Time.parse(v) if k.ends_with?("_at")
+    if v.is_a?(Array)
+      v.each do |val| 
+        if val.is_a?(Hash)
+          convert_times(val)
+        elsif val.is_a?(String)
+          val.to_datetime if val.match?(/\d{4}-\d{2}-\d{2}T/)
+        end
```
Comment (jbradl11) [inline lib/tasks/fixture_helper.rb:60]: Is there a possibility of an Array inside of an Array?

File: app/controllers/measures_controller.rb (hunk 1 above)
Comment (lukoktonos) [reply app/controllers/measures_controller.rb]: Yes, good point--the users said that they would want all dates on the patient shifted.

Comment (zlister) [reply app/controllers/measures_controller.rb]: Is the death Datetime a data element? It might already be handled then

File: lib/tasks/fixture_helper.rb (hunk 3 above)
Comment (zlister) [reply lib/tasks/fixture_helper.rb:60]: I haven't seen one, usually it is an array of Hashes (an array containing other classes tend to be a Hash). I can add another check just so that it is covered in case though

File: app/controllers/measures_controller.rb (hunk 1 above)
Comment (zlister) [reply app/controllers/measures_controller.rb]: `expiredDatetime` is a data_element

**Review by lukoktonos — COMMENTED** [review]
It looks like CQM Patients' QDM Patients have both a `dataElements` array as well as `extendedData.source_data_criteria`.  I'm not sure what the role of the latter is--should that even be on the qdm patient?  If so, I think we may need to adjust those dates as well, as they have start/end dates.

It does not look like the qdmpatient's birthdate is being updated with the change in measurementperiod.  (the one in dataElements is, but not the other one.  The date picker in the patient builder view doesnt show a different birthdate)

It would be better for the error upon failure to shift an element (due to <1 or >9999) to give some information about which element is causing the issue if possible:
```Element on Patient could not be shifted. Please make sure shift will keep all years between 1 and 9999```

I think it may make sense to hold off on merging this PR until the patient saving and patient builder timing section get in, because those allow for more comprehensive manual testing.

File: Gemfile.lock [hunk 4]
```diff
@@ -49,7 +49,7 @@ GIT
 
 GIT
   remote: https://github.com/projecttacoma/cqm-models
-  revision: 2688b71f537e97d4b9b126f453816b0c7fffda60
+  revision: ef59618fa137ba6fa27c057b2a57cea4977ba0fa
```
Comment (lukoktonos) [inline Gemfile.lock]: `2688b71` is the most up-to-date, `ef59618f` is 1 commit behind now

File: lib/tasks/fixture_helper.rb [hunk 5]
```diff
@@ -45,12 +45,36 @@ def load_fixture_file(file, collection_name, user = nil)
...
+# Some date fields can be nested so that is why there is recursion
 def convert_times(json)
   return nil unless json.is_a?(Hash)
   json.each_pair do |k,v|
-    json[k] = Time.parse(v) if k.ends_with?("_at")
+    if v.is_a?(Array)
+      v.each do |val|
+        if val.is_a?(Hash)
+          vals = [val]
+          vals.each { |vall| convert_times(vall) }
```
Comment (jbradl11) [inline lib/tasks/fixture_helper.rb]: You should be able to just to do this
```
if val.is_a?(Hash)
  convert_times(val)
```
Since you have a base check for `Hash` this will be handled.

File: lib/tasks/fixture_helper.rb [hunk 6]
```diff
@@ -45,12 +45,36 @@ def load_fixture_file(file, collection_name, user = nil)
...
+        elsif val.is_a?(String)
+          val.to_datetime if val.match?(/\d{4}-\d{2}-\d{2}T/)
+        elsif val.is_a?(Array)
+          val.each do |value|
```
Comment (jbradl11) [inline lib/tasks/fixture_helper.rb]: Similarly to the previous comment. Since you have a base check for `Array` you can just pass val back into the function.

**Review by lukoktonos — APPROVED** [review]
Looks good!

**Review by jbradl11 — APPROVED** [review]
